### PR TITLE
Remove dq.rand namespace

### PR DESCRIPTION
**Pull request by gautierronan**
Also moved `dq.rand_real` and `dq.rand_complex` to the new `utils/random.py` file, and removed the duplicate `dq.rand_complex` function.

Left to do in this PR: 
- [x] add documentation for other `dq.rand_*` functions.

Related to https://linear.app/dynamiqs/issue/DYN-21/improve-random-module.

### CODE REVIEW

**Review by abocquet — DISMISSED** [review]

File: docs/python_api/index.md [hunk 1]
```diff
@@ -165,11 +165,18 @@ The **dynamiqs** Python API features two main types of functions: solvers of dif
...
         - cd_gate
 
+### Random arrays
+
+::: dynamiqs.utils.random
+    options:
+        table: true
+        members:
+        - rand_real
+        - rand_complex
```
Comment (pierreguilmin) [inline docs/python_api/index.md:178]: We are missing the new guys here 😋

File: dynamiqs/utils/random.py [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+    """
+    if not len(shape) >= 2 or not shape[-1] == shape[-2]:
+        raise ValueError(f'`shape` must be at least 2D and square, but got {shape}.')
+    x = rand_complex(key, shape)
+    return x @ dag(x)
+
+
+def rand_dm(key: PRNGKeyArray, shape: tuple[int, ...]) -> Array:
+    """Returns a random density matrix (positive semi-definite, hermitian and unit
+    trace).
```
Comment (pierreguilmin) [inline dynamiqs/utils/random.py:146]:
```suggestion
    """Returns a random density matrix (hermitian, positive semi-definite, and unit
    trace).
```

(positive semi-definite => hermitian, so I prefer this order)

File: dynamiqs/utils/random.py [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+               [ 0.473+0.446j,  0.13 +0.j   ]], dtype=complex64)
+    """
+    if not len(shape) >= 2 or not shape[-1] == shape[-2]:
+        raise ValueError(f'`shape` must be at least 2D and square, but got {shape}.')
+    x = rand_complex(key, shape)
+    return 0.5 * (x + dag(x))
+
+
+def rand_psd(key: PRNGKeyArray, shape: tuple[int, ...]) -> Array:
+    """Returns a random complex positive semi-definite array.
```
Comment (pierreguilmin) [inline dynamiqs/utils/random.py:122]:
```suggestion
    """Returns a random complex positive semi-definite matrix.
```

File: dynamiqs/utils/random.py [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+               [-4.428+1.744j, -0.53 +1.668j,  2.582+0.65j ]], dtype=complex64)
+    """
+    shape = (shape,) if isinstance(shape, int) else shape
+    # sample uniformly in the unit L2 ball and scale
+    x = rmax * jax.random.ball(key, 2, shape=shape)
+    return x[..., 0] + 1j * x[..., 1]
+
+
+def rand_herm(key: PRNGKeyArray, shape: tuple[int, ...]) -> Array:
+    """Returns a random complex Hermitian array.
```
Comment (pierreguilmin) [inline dynamiqs/utils/random.py:100]:
```suggestion
    """Returns a random complex Hermitian matrix.
```

**Review by pierreguilmin — APPROVED** [review]
Great thanks! 👏

**Review by pierreguilmin — APPROVED** [review]

**Pull request merged**